Test/dafny0/NoMoreAssume2Less2.dfy

```diff
@@ -0,0 +1,199 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// This file tests various ways of writing the equivalent of "assume false;" or "assert false;" inside a loop.
+// It used to be that Dafny's translation of such statements into Boogie could cause Boogie to cut off the
+// loop back edge, which would turn a loop into something that wasn't a loop. In particular, it would cause
+// the expected "havoc loop_targets;" statement never to be generated, whose dramatic impact on what gets
+// verified would be both mysterious and surprising to most Dafny users.
+//
+// The equally mysterious workaround for this problem used to be for Dafny users to replace their "assume false;"
+// with something like "assume 2 < 2;", which Boogie wouldn't recognize in the same way.
+//
+// The current translation to Boogie makes sure that no user-supplied expression turns into exactly an
+// "assume false;" or "assert false;". This means that Dafny users can use a statement like "assume false;"
+// without chopping off back edges.
+//
+// It is still possible (in situations that would be rare in practice) to get some initially-surprising behavior
+// from Boogie, because it is possible to introduce a "false" condition that will be detected by the control-flow
+// handling in Boogie's abstract interpreter. Method Conjuncts4 below shows such an example.
+
+method M0(n: nat) {
+  var i := 0;
+  while i < n {
+    i := i + 1;
+  }
+  assert n == 0; // error
+}
+
+method M1(n: nat) {
+  var i := 0;
+  while i < n {
+    assume false;
+    i := i + 1;
+  }
+  assert n == 0; // error
```

Comment:
Why did this line not produce an error before this PR? 

Even if the `while` is seen as something that's not a loop, like an `if`, line 35 is still reachable since the expression `i < n` can be false right away, skipping the entire while, and thus line 35 should have given a verification error, right?

Could you help me understand what Boogie is doing here that causes the unexpected behavior?

The explanation `In particular, it would cause the expected "havoc loop_targets;" statement never to be generated, whose dramatic impact on what gets verified would be both mysterious and surprising to most Dafny users.` doesn't seem to relate to this test-case.

Reply:
Early in the Boogie pipeline, Boogie does two things:
* It desugars `while` statements into a bunch of basic blocks and `goto` statements. Note: Once this is done, cycles between basic blocks still remain in the control-flow graph, but there are no longer any structured `while` statements. (Because of the `goto` statement, the original Boogie program may also have contained loops that didn't use `while`.) Among the resulting basic blocks, it is still possible to _compute_ where there are loops, by traversing through the basic blocks and figuring out which blocks are loop heads.
* It looks for statements of the syntactic form `assume false;` or `assert false;`, and then chops off any control flow that follows. After all, there is no flow of control that passes through an `assume false;` or `assert false;` and gets to the other side.

Boogie does both of these before it computes which blocks are loops heads. If an `assume false;` is placed before what would have been a back edge (that is, the `}` at the end of a `while` body, or for that matter a `goto` back to the loop head), then there is no longer a loop for Boogie loop-head reconstruction to find.

You are right that, even if the `while` on line 31 in effect turns into a `if`, there still is a control flow from the "else" branch that leads to line 35. But that control can take place only if `i < n` is `false` initially, which implies that `n == 0`. So the `assert n == 0;` on line 35 passes.

When I said `... "havoc loop_targets;" ...`, I was alluding to something that Boogie does when it deals with loops. What's relevant here is that an `assume false;` can cause Boogie to not find any loops.

Reply:
Just for context, I read your "Weakest-Precondition of Unstructured Programs" which I think mostly explains what Boogie does.

> But that control can take place only if i < n is false initially, which implies that n == 0. So the assert n == 0; on line 35 passes.

Oops, I missed that.

Does the `havoc i` at the start of the loop cause the `var i := 0` information to become lost, and that's why it produces a verification error now?